kratos/elements/distance_calculation_flux_based_element.cpp

```diff
@@ -0,0 +1,502 @@
+        nodal_values[i] = GetGeometry()[i].FastGetSolutionStepValue(DISTANCE);
+    }
+
+    //computing density so that Fourier number of the complete domain = 1
+    //this ensures that all regions notice the effect of the fixed nodes.
+    const double domain_length = rCurrentProcessInfo[CHARACTERISTIC_LENGTH];
+    const double density  = 1.0/ (domain_length*domain_length);
+
+    
+    //Looping Gauss Points
+    for (unsigned int igauss = 0; igauss<NumNodes; igauss++) {
+        array_1d<double, TNumNodes > N = row(N_container, igauss);
+        noalias(rLeftHandSideMatrix) += gauss_weights[igauss]*prod(DN_DX_container[igauss],trans(DN_DX_container[igauss]));
+        const double mass_factor = density*gauss_weights[igauss];
+        const double d_gauss = inner_prod(N,nodal_values);
+        const double initial_value = d_gauss > 0.0 ? domain_length*10.0 : -domain_length*10.0 ;
+        for(unsigned int j=0; j<NumNodes; j++){
+            rLeftHandSideMatrix(j,j)+= mass_factor*N[j];
+            rRightHandSideVector[j] += mass_factor*N[j]*initial_value;
+        }
+    }
+    
+    //substracting previous solution
+    noalias(rRightHandSideVector) -= prod(rLeftHandSideMatrix,nodal_values);
+}
+
+    
+template <unsigned int TDim, unsigned int TNumNodes >
+void DistanceCalculationFluxBasedElement<TDim, TNumNodes >::CalculateDistanceSystem(
+        MatrixType &rLeftHandSideMatrix,
+        VectorType &rRightHandSideVector,
+        const ProcessInfo &rCurrentProcessInfo)
+{
+    BoundedVector<double,TNumNodes> gauss_weights;
+    BoundedMatrix<double,TNumNodes,TNumNodes> N_container;
+    array_1d<BoundedMatrix<double, TNumNodes, TDim>, TNumNodes> DN_DX_container;
+
+    CalculateGaussPointsData( GetGeometry(), gauss_weights, N_container, DN_DX_container);
+
+    array_1d<double, NumNodes > nodal_values;
+    array_1d< array_1d<double, TDim >, NumNodes> v; //convection velocity
+    array_1d< array_1d<double, TDim >, NumNodes> v_unit; //to decide if convection must be turned off
+    array_1d<double, TDim > avg_v_unit = ZeroVector(TDim); //to decide if convection must be turned off
+    bool has_fixed_node = false;
+
+    for(unsigned int i=0; i<NumNodes; i++){
+        nodal_values[i] = GetGeometry()[i].FastGetSolutionStepValue(DISTANCE);
+        const auto& rVel_i = GetGeometry()[i].GetValue(POTENTIAL_GRADIENT);
+        for(unsigned int j=0; j<TDim; j++){
+            v[i][j] = rVel_i[j];
+        }
+        v_unit[i] =  v[i]/norm_2(v[i]); // |v_unit| = 1
+        avg_v_unit += v_unit[i];
+        has_fixed_node = has_fixed_node || GetGeometry()[i].IsFixed(DISTANCE);
+    }
+    avg_v_unit/=static_cast<double>(NumNodes);
+
+    BoundedMatrix<double, TNumNodes, TDim> avg_DN_DX = ZeroMatrix(TNumNodes, TDim);
+    for(unsigned int i=0; i<TNumNodes; i++){
+        avg_DN_DX += DN_DX_container[i] /static_cast<double>(TNumNodes);
+    }
+    
+    //adjusting RHS to get |grad|=1
+    if(rCurrentProcessInfo[NL_ITERATION_NUMBER] > 1) {
```

Comment:
i've just removed the correction coefficient.
basically, it is not possible to guarantee grad=1 everywhere when using the diffusion equation to define the direction of the gradient of the distance. Using this equations means that from the source (dist=0) there will be diverging and converging "streamlines" . But this is incorrect, since a converging streamline will have an inconsistent value on what the distance should be at that converging point. The best approach i've found is to use the gradient of the diffusion eq as a weight. the higher it is, the shorter the path and hence the closest to the real distance.
so to sum up, this distance method is not exact, but a more robust approximation